nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1377,41 +1362,80 @@ impl DataStore {
-                    match instance_state {
-                        api::external::InstanceState::Stopped
-                        | api::external::InstanceState::Failed => {
-                            Error::internal_error("cannot delete instance")
-                        }
-                        _ => Error::invalid_request(&format!(
-                            "instance cannot be deleted in state \"{}\"",
-                            instance_state,
-                        )),
-                    }
+                    ));
                 }
-                DetachManyError::DatabaseError(e) => {
-                    public_error_from_diesel(e, ErrorHandler::Server)
+                let instance_state =
+                    collection.runtime_state.nexus_state.state();
+                match instance_state {
+                    api::external::InstanceState::Stopped
+                    | api::external::InstanceState::Failed => {
+                        Err(Error::internal_error("cannot delete instance"))
+                    }
+                    _ => Err(Error::invalid_request(&format!(
+                        "instance cannot be deleted in state \"{}\"",
+                        instance_state,
+                    ))),
                 }
-            })?;
+            }
+            DetachManyError::DatabaseError(e) => {
+                Err(public_error_from_diesel(e, ErrorHandler::Server))
+            }
+        })?;
+
+        Ok(())
+    }
 
+    /// Delete the provided `authz_instance`, as long as it is in one of the
+    /// provided set of [`InstanceState`]s.
+    ///
+    /// A.K.A. "[`project_delete_instance`], hard mode". Typically, instances
+    /// may only be deleted if they are in the [`InstanceState::NoVmm`] or
+    /// `[`InstanceState::Failed`] states. This method exists separately in
+    /// order to allow an unwinding `instance-create` saga to delete the
+    /// instance record without transiently moving it to
+    /// [`InstanceState::Failed`], in which it becomes eligible for automatic
+    /// restart.
+    ///
+    /// In general, callers should use [`project_delete_instance`] instead,
+    /// unless you really know what you're doing.
+    ///
+    /// [`project_delete_instance`]: Self::project_delete_instance
+    pub async fn project_delete_instance_in_states(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        ok_to_delete_instance_states: &'static [InstanceState],
+    ) -> DeleteResult {
+        // First, mark the instance record as deleted
+        self.project_idempotently_mark_instance_deleted_in_states(
+            opctx,
+            authz_instance,
+            ok_to_delete_instance_states,
+        )
+        .await?;
+
+        // Next, delete all other data associated with the instance.
+        //
+        // Note that due to idempotency of this function, it's possible that
+        // "authz_instance.id()" has already been deleted.
```

Comment:
...but that's fine, since we don't use foreign key constraints?

Reply:
Yup. It's possible that these functions either "match on a row with a `.filter(instance_id.eq(id))`, or, they don't"